pkg/util/token_bucket.go

```diff
@@ -0,0 +1,95 @@
+package util
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type TokenBucket struct {
+	remainingTokens int64
+	maxCapacity     int64
+	refillRate      int64
+	lastRefill      time.Time
+	mu              sync.Mutex
+
+	remainingTokensMetric prometheus.Gauge
```

Comment:
Wouldn't a gauge change a lot between each scrape? I haven't fully thought this through - Could we use a counter for allowed tokens, throttled tokens and total tokens instead?

Reply:
Purpose of the metric was to determine the remaining capacity of the token bucket, which wouldn't be achievable with count. The other option is just to have a count metric for throttled requests + retrieved tokens, but not sure if we'll make use of the retrieved token metric then